### PR TITLE
All object payloads should have an UID

### PR DESCRIPTION
**Pull request by leonorfmartins**
**What is this feature?**

This does several things:
- adds asyncronicity to Mode 3
- adds a condition on apiserver to only accept requests with a populated UID
- simplifies the dual writer layer A LOT after https://github.com/grafana/grafana/pull/93714 where we delegate to each resource to provide the needed payload for the dual writer to work.

**Next up:**
Uncomment 
```
	 if obj.GetResourceVersion() != "" {
	 	return nil, NewBadRequestError("object must not include resourceVersion")
	 }
```
and fix the rest of the tests.

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Part of https://github.com/grafana/search-and-storage-team/issues/96

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

**Comment by leonorfmartins** [comment]
Part of https://github.com/grafana/search-and-storage-team/issues/96

File: pkg/apiserver/rest/dualwriter_mode2.go [hunk 1]
```diff
@@ -48,39 +48,44 @@ func (d *DualWriterMode2) Mode() DualWriterMode {
...
 		log.Error(err, "unable to create object in legacy storage")
 		d.recordLegacyDuration(true, mode2Str, d.resource, method, startLegacy)
-		return created, err
+		return createdFromLegacy, err
 	}
 	d.recordLegacyDuration(false, mode2Str, d.resource, method, startLegacy)
 
-	if err := enrichLegacyObject(original, created); err != nil {
-		return created, err
+	accIn, err := meta.Accessor(in)
```
Comment (toddtreece) [inline pkg/apiserver/rest/dualwriter_mode2.go:65]: i think this check could be removed. the underlying storage layer should validate this

File: pkg/apiserver/rest/dualwriter_mode3.go [hunk 2]
```diff
@@ -42,30 +44,57 @@ func (d *DualWriterMode3) Mode() DualWriterMode {
...
 	}
-	d.recordStorageDuration(false, mode3Str, d.resource, method, startStorage)
 
-	go func() {
-		ctx, cancel := context.WithTimeoutCause(ctx, time.Second*10, errors.New("legacy create timeout"))
-		defer cancel()
+	createdCopy := storageObj.DeepCopyObject()
+
+	//nolint:errcheck
+	go d.createOnLegacyStorage(ctx, createdCopy, createValidation, options)
```
Comment (toddtreece) [inline pkg/apiserver/rest/dualwriter_mode3.go:73]: what if we need to roll back to legacy? is there a reverse sync?

File: pkg/apiserver/storage/testing/watcher_tests.go [hunk 3]
```diff
@@ -44,11 +44,11 @@ func RunTestWatch(ctx context.Context, t *testing.T, store storage.Interface) {
 // - update that gets filtered should trigger Deleted event
 func testWatch(ctx context.Context, t *testing.T, store storage.Interface, recursive bool) {
 	basePod := &example.Pod{
-		ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+		ObjectMeta: metav1.ObjectMeta{Name: "foo", UID: "123"},
```
Comment (toddtreece) [inline pkg/apiserver/storage/testing/watcher_tests.go:47]: i don't think these should need to be modified to pass 🤔

File: pkg/apiserver/rest/dualwriter_mode2.go (hunk 1 above)
Comment (leonorfmartins) [reply pkg/apiserver/rest/dualwriter_mode2.go:65]: I'd rather keep returning the error here as it's easier for someone without the whole context or know from where this is called to understand

File: pkg/apiserver/rest/dualwriter_mode3.go (hunk 2 above)
Comment (leonorfmartins) [reply pkg/apiserver/rest/dualwriter_mode3.go:73]: there aren't plans for that, but it's actually something we want to discuss. in the team: if there is one, that will mean that mode 3 will be pretty similar to mode 2 :/

File: pkg/apiserver/rest/dualwriter_mode1.go [hunk 4]
```diff
@@ -63,22 +63,28 @@ func (d *DualWriterMode1) Create(ctx context.Context, original runtime.Object, c
...
 
-	if err := enrichLegacyObject(original, createdCopy); err != nil {
-		cancel()
+	accIn, err := meta.Accessor(in)
+	if err != nil {
+		return err
+	}
+
+	if accIn.GetUID() != "" {
+		return fmt.Errorf("there is an UID and it should not: %v", accIn.GetUID())
```
Comment (inf0rmer) [inline pkg/apiserver/rest/dualwriter_mode1.go:80]:
```suggestion
		return fmt.Errorf("a UID is present, but it should not be: %v", accIn.GetUID())
```

File: pkg/apiserver/storage/testing/watcher_tests.go (hunk 3 above)
Comment (leonorfmartins) [reply pkg/apiserver/storage/testing/watcher_tests.go:47]: they actually fail without it 😞

**Pull request closed without merging**